components/script/dom/mutationobserver.rs

```diff
@@ -37,4 +70,163 @@ impl MutationObserver {
+        let mut pairedStrings: Vec<DOMString> = vec![];
+        // Step 2
+        let mut nodes: Vec<Root<Node>> = vec![];
+        for ancestor in target.inclusive_ancestors() {
+            nodes.push(ancestor);
+        }
+        // Step 3
+        for node in &nodes {
+            for registered_observer in node.registered_mutation_observers().borrow().iter() {
+                match attr_type {
+                    // TODO check for Mutations other than Attribute
+                    Mutation::Attribute { ref name, ref namespace, ref oldValue, ref newValue } => {
+                        let condition1: bool = node != &Root::from_ref(target) &&
+                            !registered_observer.subtree.get();
+                        let condition2: bool = registered_observer.attributes.get() == false;
+                        let condition3: bool = !registered_observer.attribute_filter.borrow().is_empty(); &&
+                            (registered_observer.attribute_filter.borrow().iter()
+                                .find(|s| &**s == &**name).is_none() || namespace != &ns!());
+                        if !condition1 && !condition2 && !condition3 {
+                            // Step 3.1.2
+                            let paired_string = if registered_observer.attribute_old_value.get() {
+                                DOMString::from(oldValue.deref())
+                            } else {
+                                DOMString::from("")
+                            };
+                            // Step 3.1.1
+                            let idx = interestedObservers.iter().position(|&(ref o, _)|
+                                &**o as *const MutationObserver == &**registered_observer as *const MutationObserver);
+                            if let Some(idx) = idx {
+                                interestedObservers[idx].1 = paired_string;
+                            } else {
+                                interestedObservers.push((Root::from_ref(registered_observer), paired_string));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        // Step 4
+//        let mut i = 0;
+        for (i, observer) in interestedObservers.iter().enumerate() {
+            //Step 4.1
+            let record = &MutationRecord::new(DOMString::from("attributes"), target);
+            //Step 4.2
+            match attr_type {
+                Mutation::Attribute { ref name, ref namespace, ref oldValue, ref newValue } => {
+                    record.SetAttributeName(DOMString::from(&**name));
+                    record.SetAttributeNamespace(DOMString::from(&**namespace));
+                }
+            }
+            //Step 4.3-4.6- TODO currently not relevant to mutation records for attribute mutations
+            //Step 4.7
+            if pairedStrings[i] != "" {
+                record.SetoldValue(pairedStrings[i].clone());
```

Comment:
This will need to use the `paired_string` variable from the iterator instead. The `pairedStrings` vector should be removed to avoid accidentally using it.